src/passes/ConstantFieldPropagation.cpp

```diff
@@ -0,0 +1,450 @@
+    if (!info.isConstant()) {
+      return;
+    }
+
+    // We can do this! Replace the get with a trap on a null reference using a
+    // ref.as_non_null (we need to trap as the get would have done so), plus the
+    // constant value. (Leave it to further optimizations to get rid of the
+    // ref.)
+    replaceCurrent(builder.makeSequence(
+      builder.makeDrop(builder.makeRefAs(RefAsNonNull, curr->ref)),
+      builder.makeConstantExpression(info.getConstantValue())));
+    changed = true;
+  }
+
+  void doWalkFunction(Function* func) {
+    WalkerPass<PostWalker<FunctionOptimizer>>::doWalkFunction(func);
+
+    // If we changed anything, we need to update parent types as types may have
+    // changed.
+    if (changed) {
+      ReFinalize().walkFunctionInModule(func, getModule());
+    }
+  }
+
+private:
+  StructValuesMap& infos;
+
+  bool changed = false;
+};
+
+struct ConstantFieldPropagation : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "ConstantFieldPropagation requires nominal typing";
+    }
+
+    // Find and analyze all writes inside each function.
+    FunctionStructValuesMap functionInfos;
+    for (auto& func : module->functions) {
+      // Initialize the data for each function, so that we can operate on this
+      // structure in parallel without modifying it.
+      functionInfos[func.get()];
+    }
+    Scanner scanner(functionInfos);
+    scanner.run(runner, module);
+    scanner.walkModuleCode(module);
+
+    // Combine the data from the functions.
+    StructValuesMap combinedInfos;
+    for (auto& kv : functionInfos) {
+      StructValuesMap& infos = kv.second;
+      for (auto& kv : infos) {
+        auto type = kv.first;
+        auto& info = kv.second;
+        for (Index i = 0; i < info.size(); i++) {
+          combinedInfos[type][i].combine(info[i]);
+        }
+      }
+    }
+
+    // Handle subtyping. |combinedInfo| so far contains data that represents
+    // each struct.new and struct.set's operation on the struct type used in
+    // that instruction. That is, if we do a struct.set to type T, the value was
+    // noted for type T. But our actual goal is to answer questions about
+    // struct.gets. Specifically, when later we see:
+    //
+    //  (struct.get $A x (REF-1))
+    //
+    // Then we want to be aware of all the relevant struct.sets, that is, the
+    // sets that can write data that this get reads. Given a set
+    //
+    //  (struct.set $B x (REF-2) (..value..))
+    //
+    // then
+    //
+    //  1. If $B is a subtype of $A, it is relevant: the actual value might be
+    //     of type $B (or even more specific), and cast to $A before it reaches
+    //     the get.
+    //  2. If $B is a supertype of $A that still has the field x then it may
+    //     also be relevant: the actual value might be of type $A (or even more
+    //     specific), and cast to $B before it reaches the set.
```

Comment:
I was confused by this at first, but I get it now. Here's how I would explain it:

Given (struct.get $A x ...) and (struct.set $B x ...)

 - If $A <: $B, then (struct.set $B x ...) can write a struct of type $A.
 - If $B <: $A, then (struct.get $A x ...) can read a struct of type $B.

In both cases, the struct.get can observe the struct.set, so struct.sets to all subtypes and supertype must agree for a get of a type to be optimized.

As you've noted, struct.new does not have this bidirectionality and could be modeled separately. This will be important for devirtualization because vtables are never written to after their creation and also because vtable subtypes will have different values than their supertypes.

Reply:
Good summary, updated the comment now.